src/tools/crossgen2/ILCompiler.DependencyAnalysisFramework/PerfEventSource.cs

```diff
@@ -10,22 +11,49 @@
 namespace ILCompiler.DependencyAnalysisFramework
 {
     // The event IDs here must not collide with the ones used by ReadyToRunPerfEventSource.cs
-    [EventSource(Name = "Microsoft-ILCompiler-Perf")]
-    class PerfEventSource : EventSource
+    struct StartStopEvents : IDisposable
     {
-        [Event(1001, Level = EventLevel.Informational)]
-        public void GraphProcessingStart() { WriteEvent(1001); }
-        [Event(1002, Level = EventLevel.Informational)]
-        public void GraphProcessingStop() { WriteEvent(1002); }
+        private Action _stopAction;
+        public void Dispose()
+        {
+            _stopAction();
+        }
 
-        [Event(1003, Level = EventLevel.Informational)]
-        public void DependencyAnalysisStart() { WriteEvent(1003); }
-        [Event(1004, Level = EventLevel.Informational)]
-        public void DependencyAnalysisStop() { WriteEvent(1004); }
+        private StartStopEvents(Action startAction, Action stopAction)
+        {
+            startAction();
+            _stopAction = stopAction;
+        }
 
-        [Event(1005, Level = EventLevel.Informational)]
-        public void AddedNodeToMarkStack() { WriteEvent(1005); }
+        [EventSource(Name = "Microsoft-ILCompiler-Perf")]
+        public class PerfEventSource : EventSource
+        {
+            public static PerfEventSource Log = new PerfEventSource();
 
-        public static PerfEventSource Log = new PerfEventSource();
+            private PerfEventSource() {}
+
+            public StartStopEvents GraphProcessingEvents()
+            {
+                return new StartStopEvents(GraphProcessingStart, GraphProcessingStop);
```

Comment:
This allocates two new delegates on each call. For some of these events this doesn't matter because they're only called a couple times. But for others this is more important. We should have an efficient precedent that every current and future event follows.

* Maybe we could just call `GraphProcessingStart` right here instead of creating a delegate that we then invoke in the `StartStopEvent` constructor and throw away.
* If `EventSource.IsEnabled` is false (nobody enabled this provider), we shouldn't create the delegate at all and pass `null` to the `StartStopEvents` constructor.

Reply:
Oh right, I forgot that delegates aren't just function pointers and actually have some overhead. Changed.